Support multiple elk instances in one process.

gjbadros opened the pull request:
This moves _message_handler from being file-local in messages.py to being a field in the Elk object.

Pipfile and Pipfile.lock changes seem spurious, but those weren't listed in .gitignore, so I left the diffs.

gwww commented:
Thank you! Not what I was expecting. I was thinking you were talking about the Elkm1 library. No worries though. Since this is a rather substantial PR could you post on the discord channel (there's a long ElkM1 thread there) and let people know to see if they will try it out. I'd like to get more eyes on it than just mine, and, I won't be able to properly test this for a while.

gwww commented:
I'll take a look.

gjbadros commented:
It's both a change to the elkm1 library and a change to hass's elkm1
component.  The pr was for the elkm1 library.  Is that what you want posted
on discord?

It's actually not a big change in principle... Just touches all the files
with a minor change.


On Wed, Mar 27, 2019, 2:34 PM Glenn Waters <notifications@github.com> wrote:

> Thank you! Not what I was expecting. I was thinking you were talking about
> the Elkm1 library. No worries though. Since this is a rather substantial PR
> could you post on the discord channel (there's a long ElkM1 thread there)
> and let people know to see if they will try it out. I'd like to get more
> eyes on it than just mine, and, I won't be able to properly test this for a
> while.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gwww/elkm1/pull/5#issuecomment-477356396>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbjzQW3pxHInn09eqiU-6aPfGb43znxks5va-PUgaJpZM4cOjUn>
> .
>

gwww commented:
Overall looks pretty good. Thank you for the changes to support multiple Elk panels!

I was confused because the patch was for hass then the pr was for the lib. I now see you have changed both. My comment about discord is for the hass change, although letting people know that the lib is changing is also a good thing.

I think that the add_message_handler should be a method on Elk not on Element. The _message_handlers resides with Elk and should be manipulated from there. Does that sound right to you?

gjbadros commented:
The methods on Element and Elements are just helper functions -- are you
saying to move message_handler out of the message.py?  I was trying to
avoid message.py taking a dependency on elk.py's Elk object, and as such,
add_message_handler staying in the code with the rest of the implementation
of message handling made some sense and was a more minimal change.

I initially considered just moving all of message.py into elk.py, but that
seemed like a bigger change so went with this simpler approach.

Greg

On Wed, Mar 27, 2019 at 2:52 PM Glenn Waters <notifications@github.com>
wrote:

> Overall looks pretty good. Thank you for the changes to support multiple
> Elk panels!
>
> I was confused because the patch was for hass then the pr was for the lib.
> I now see you have changed both. My comment about discord is for the hass
> change, although letting people know that the lib is changing is also a
> good thing.
>
> I think that the add_message_handler should be a method on Elk not on
> Element. The _message_handlers resides with Elk and should be manipulated
> from there. Does that sound right to you?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gwww/elkm1/pull/5#issuecomment-477361556>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbjzXMo6vru3X7xWWMY8T_LOOrjf6rIks5va-gWgaJpZM4cOjUn>
> .
>

gwww commented:
What I was suggesting is add `add_message_handler` method in `Elk` class.

I do believe is that the `message_handlers` array should be in `Elk`, as you have changed. 

What is causing me to wonder is having a "helper" in `Element`; it feels out of place. I'm OK with leaving `add_message_handler` in `message.py` as a helper. I view all of `message.py` to be a util/helper file.

Does that make sense?

gjbadros commented:
How's this look now?  I kept the helpers in Element and Elements, but they're simply forwarding functions.  add_message_handler helper is still in message.py and elk.py has an _add_message_handler that calls that helper and is forwarded to by Element and Elements.

gjbadros commented:
Just noticed a bug -- there's a couple more file globals in util.py -- stay tuned for an update to the PR. It was manifesting itself by not all the elk's updating their zone names (I had thought that was a bug in my hass implementation that I've not PR'd yet, but it's an omission in my library improvements)

gwww commented:
Cool. Love what you're doing!

I think that I've figured out what was bugging me about the message
handling. Although I stand to still be wrong.

It's that the message handlers array is passed around and accessed in
multiple places.

Please be patient as we try to get it "right".

Currently thinking that the code in message.py should be a class. An
instance of the class is created by the main Elk class. This would
encapsulate all the message handler code. Perhaps the decoders too.

Does that make sense?

So you know, when I wrote the library, some parts were refactored a dozen
times until they felt right. 🤔

As you can see, I'm typing on my phone, so hard to be too much more
expressive at the moment.

Cheers, Glenn
______
Sent using an itsy bitsy keyboard. Expect typos and brevity.

On Thu, Mar 28, 2019, 16:46 gjbadros, <notifications@github.com> wrote:

> Just noticed a bug -- there's a couple more file globals in util.py --
> stay tuned for an update to the PR. It was manifesting itself by not all
> the elk's updating their zone names (I had thought that was a bug in my
> hass implementation that I've not PR'd yet, but it's an omission in my
> library improvements)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gwww/elkm1/pull/5#issuecomment-477764517>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC6xoHh0waVHwYwVfKgbOd3t316n4ACmks5vbSoqgaJpZM4cOjUn>
> .
>

gjbadros commented:
Have another look please -- I've looked for other file-global objects and I'm pretty confident that the ones in util.py are the last that I missed.

And yea, I agree with you that putting some of the messages code in a class might make it a bit cleaner.  I can imagine a semantics-preserving rewrite that does that with a proper registration pattern (more like what you'd do in a static-typed language without using globals() and lexical creation of the function name to invoke) but that's a bigger rewrite that makes it less obvious that the change is correct.  It also might be unnecessary given how complete this code otherwise is.  My view is that with the refactoring to simplify that I implemented, the pay-off for encapsulating messages is not there and the importance of fixing the Elk singleton assumption is high.

Re: the test cases for get_descriptions, etc., I'm not sure there's a great way to do those and the test coverage of the class hierarchy is virtually zero, so now that that functionality lives inside Elk and Elements it is not out of place to not have good test coverage.  (That written, I've never used this Mock() class before so maybe I'm missing something about how it could be done painlessly.)

gwww commented:
Take a look at the new-eh branch. It's not perfect (timeout and unknown messages not working properly). Thoughts?

gjbadros commented:
Cool! I don't love the name "Message" for the class -- maybe
MessageProcessor (since Handler is used for something else).  Message makes
me think the instances will be, well, messages.

Is the intent for this to replace my CL or just provide an example of what
the "Message" class could look like?  If the former, there's still file
globals in util.py related to sync handlers and description processing that
need to be moved into elk.

If the latter, it might be nice to have a dictionary of all the allowed
commands to validate the command before trying to look it up.  I suppose
it's not too risky given the extra structure on the method name
(/_.._decode/), but I think getattr() chases up the inheritance tree so it
could be non obvious that there's a surprise method that could be invoked
by a nefarious Elk-pretending server.

Do you know why unknown messages and timeouts don't work?



On Fri, Mar 29, 2019 at 4:58 PM Glenn Waters <notifications@github.com>
wrote:

> Take a look at the new-eh branch. It's not perfect (timeout and unknown
> messages not working properly). Thoughts?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gwww/elkm1/pull/5#issuecomment-478183383>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbjzWpy5xwTHPb1vRYwyicZAEZSlFgAks5vbqidgaJpZM4cOjUn>
> .
>

gwww commented:
Thanks for the feedback! Yah, `Message` is a bit weak. Can change. I like the way the class came together. It feels right, with the handlers stored in the Elk class and the the message.py nicely holds all the message stuff.

I know why timeout and unknown don't work. Just need more refactoring.

How about this, I add this change for the message handling, you submit a PR for the stuff in util.py?  Or... I'm open to suggestions! I wasn't really planning on getting into this but couldn't resist a nice design problem!

gjbadros commented:
I probably can just pull down your message.py change and then almost
everything else in my fork will work... That would be easier for me to
manage from a change perspective. Or remake your message class changes on
my fork?  I think I gave you edit permission. Then do the pull into the
main repo.

On Fri, Mar 29, 2019, 5:53 PM Glenn Waters <notifications@github.com> wrote:

> Thanks for the feedback! Yah, Message is a bit weak. Can change. I like
> the way the class came together. It feels right, with the handlers stored
> in the Elk class and the the message.py nicely holds all the message stuff.
>
> I know why timeout and unknown don't work. Just need more refactoring.
>
> How about this, I add this change for the message handling, you submit a
> PR for the stuff in util.py? Or... I'm open to suggestions! I wasn't really
> planning on getting into this but couldn't resist a nice design problem!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gwww/elkm1/pull/5#issuecomment-478190626>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbjzbBd2MUd1XCqCshPLd6Qpb5sxV7wks5vbrWSgaJpZM4cOjUn>
> .
>

gwww commented:
I've changed a couple of things beyond the add_handler stuff. How about I merge what I have into master, you can pull the latest, then PR against that.

I've fixed unknown. Working on timeout. Won't be hard, might affect a couple of files (elk.py and proto.py)

gjbadros commented:
Ok sounds good.

On Fri, Mar 29, 2019, 6:17 PM Glenn Waters <notifications@github.com> wrote:

> I've changed a couple of things beyond the add_handler stuff. How about I
> merge what I have into master, you can pull the latest, then PR against
> that.
>
> I've fixed unknown. Working on timeout. Won't be hard, might affect a
> couple of files (elk.py and proto.py)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gwww/elkm1/pull/5#issuecomment-478192906>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbjzaF_fPyZZT9c64YDDgS55IvkhzI8ks5vbrtDgaJpZM4cOjUn>
> .
>

gwww commented:
Master up to date with all fixes I know about; changed `Message` to `MessageDecode`

gjbadros commented:
Ok, thx; I'll try this weekend to complete my second set of changes to
avoid the single-instance elk assumption.

On Fri, Mar 29, 2019 at 6:54 PM Glenn Waters <notifications@github.com>
wrote:

> Master up to date with all fixes I know about; changed Message to
> MessageDecode
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gwww/elkm1/pull/5#issuecomment-478195590>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbjzaE6PVb7nmVmdrtfoa2PnU2o1jG8ks5vbsO4gaJpZM4cOjUn>
> .
>

gwww commented on elkm1_lib/elk.py:
```diff
@@ -25,9 +25,12 @@ def __init__(self, config, loop=None):
         self._message_decode = MessageDecode()
-
+        self._sync_handlers = []
+        self._descriptions_in_prog = {}
```
Would be nicer to use full words. Use `_descriptions_in_progress` (`prog` is ambiguous)

gwww commented on elkm1_lib/elements.py:
```diff
@@ -88,9 +91,19 @@ def _got_desc(self, descriptions):
             if descriptions[element.index] is not None:
                 element.setattr('name', descriptions[element.index], True)
 
+    def _add_message_handler(self, message_type, handler):
```
This appears to be defined twice in class. Can it be done away with altogether?

gjbadros commented:
Make those changes; see the latest I just commited.  I dropped some testing
code for the newly-in-elk sync_handler and descriptions handling... if you
have ideas on how to get that testing back, I'd love for you to accept this
pull request and make those changes, then I can test more for my system off
of your master before you do a new release.

On Sat, Mar 30, 2019 at 6:12 AM Glenn Waters <notifications@github.com>
wrote:

> *@gwww* requested changes on this pull request.
>
> Whoop! Thanks. Couple of small things to cleanup.
> ------------------------------
>
> In elkm1_lib/elk.py
> <https://github.com/gwww/elkm1/pull/5#discussion_r270624316>:
>
> > @@ -112,6 +115,34 @@ def _got_data(self, data):  # pylint: disable=no-self-use
>      def _timeout(self, msg_code):
>          self._message_decode.timeout_handler(msg_code)
>
> +    def _add_sync_handler(self, sync_handler):
>
> Methods that are available outside of a class should not start with "_"
> (private vs. public)
> ------------------------------
>
> In elkm1_lib/elk.py
> <https://github.com/gwww/elkm1/pull/5#discussion_r270624959>:
>
> > @@ -25,9 +25,12 @@ def __init__(self, config, loop=None):
>          self.connection_lost_callbk = None
>          self._connection_retry_timer = 1
>          self._message_decode = MessageDecode()
> -
> +        self._sync_handlers = []
> +        self._descriptions_in_prog = {}
>
> Would be nicer to use full words. Use _descriptions_in_progress (prog is
> ambiguous)
> ------------------------------
>
> In elkm1_lib/elements.py
> <https://github.com/gwww/elkm1/pull/5#discussion_r270625128>:
>
> > @@ -88,9 +91,19 @@ def _got_desc(self, descriptions):
>              if descriptions[element.index] is not None:
>                  element.setattr('name', descriptions[element.index], True)
>
> +    def _add_message_handler(self, message_type, handler):
>
> This appears to be defined twice in class. Can it be done away with
> altogether?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gwww/elkm1/pull/5#pullrequestreview-220819598>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABbjzVqicAdkJLdpwkmCwt6HwqDUtVWcks5vb2K2gaJpZM4cOjUn>
> .
>

gwww commented:
I bumped the version to 0.7.14; make a small tidy up

gwww commented:
PyPi updated with 0.7.14 version
